README.development.md

```diff
@@ -0,0 +1,148 @@
+
+1. If not currently connected, reconnect to the persistent session in the dev container:
+
+```sh
+$ docker exec -it triton-jp5-dev tmux -u -CC new-session -A -s dev
+```
+
+2. Build the Viam Triton module inside the container with cmake and Ninja.
+Note that the build commands here reflect the container build step for the `build` stage, adjust as necessary for your situation:
+
+```sh
+$ cd ~/opt/src/viam-mlmodelservice-triton
+$ cmake -S . -B build -G Ninja -DVIAM_MLMODELSERVICE_TRITON_TRITONSERVER_ROOT=/opt/tritonserver/tritonserver -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_PREFIX_PATH=/opt/viam
+$ cmake --build build --target install -- -v
+$ cmake --install build --prefix /opt/viam
+```
+
+Note that for NVCR containers, the `VIAM_MLMODELSERVICE_TRITON_TRITONSERVER_ROOT` is just `/opt/tritonserver`.
+
+3. Make edits as needed (either inside our outside the container) and rebuild:
+
+```sh
+$ cmake --build build --target install -- -v
+```
+
+Repeat until you arrive at a state you want to try to run.
+If you have unit tests, run them here until you are happy with the results.
+
+
+4. Reinstall the modified build to `/opt/viam`:
+```sh
+$ cmake --install build --prefix /opt/viam
+```
+
+## Deployment - Outside Container
+
+This section assumes that you already have a robot built and running on the device and using a stock copy of the Viam Triton Module.
+What we are going to do is replace the container image that the Triton Module startup script (i.e. `viam-mlmodelservice-triton.sh`) uses with a snapshot produced with the development state we currently have in our container.
+
+1. Copy the installation from `/opt/viam` in the development container into the deployment container
+
+```sh
+$ docker cp -a triton-jp5-dev:/opt/viam - | docker cp -a - triton-jp5-deploy:/opt
+```
+
+If you are using the JP5 container, you also need to copy in the Triton Server libraries.
+Note that you only need to do this step once for the deployment container.
+If you are on a JP6 or CUDA container (nvcr base), you don't need to do it at all.
+
+```sh
+$ docker cp -a triton-jp5-dev:/opt/tritonserver - | docker cp -a - triton-jp5-deploy:/opt
+```
+
+2. Find the name of the image used for the currently deployed module.
+This will likely look something like this:
+
+```sh
+$ docker image list | grep 'viamrobotics.*triton'
+ghcr.io/viamrobotics/viam-mlmodelservice-triton   0.6.0     9fc8aef5b007   5 months ago     7.71GB
+```
+
+NOTE: I'd like to eliminate this step.
+Potentially, there may be a way to mount the `/opt/viam` (and, when needed for JP5, the `/opt/tritionserver`) directories from the `-dev` container into the `-deploy` container, such that writes in the `-dev` container immediately affect the running `-deploy` container, which would then get saved via the `docker commit`.
+
+3. Overwrite that container image with a snapshot of your deployment container contents
```

Comment:
:grimacing: I really don't like this approach! It's so easy to mistakenly think you're running the production version when you're actually running a local build that you forgot about last week.

Instead, I'd add a local module and disable the production one. The local module would contain the .sh script which launches the `triton-jp6-deploy` image instead of the `ghcr.io/...` one. That way, any docker image named `ghcr.io/...` comes from the GitHub Container Repository, and any locally-built repo has an obviously-local name.

Reply:
Sure, that's a fair objection. However, I only promised an effective procedure, not an elegant one :)

I also don't personally and immediately know how to describe the steps to do the local module things vs production. So, if you do have that knowledge and would like to update this document to describe it in a follow-on review, I'll be happy to approve it.

Reply:
I'm working on it now. So far, it's not working because the docker container run by the module's .sh script seems to be a fresh, pristine copy without the extra stuff copied in. I'm relatively new to docker, so might be doing something dumb on that front.

Reply:
Okay, I've got something that works! After the `docker cp` command on line 130, run `docker commit triton-jp6-deploy triton-jp6-iterative-deploy` to turn this updated container back into an image.

Meanwhile, to compile the module itself, run `TAG=triton-jp6-iterative-deploy make -f Makefile.module viam-mlmodelservice-triton.sh` to get the shell script to use the local image name. Then, on app.viam.com, add a local module pointing to that .sh file.

Now, the only docker image that ever gets renamed is the `-iterative-deploy` one, which seems like the right approach.

Reply:
Great. Let's merge this, and you can amend it with your improved process. I don't think it makes sense for me to write a step by step guide for something you have practice doing.